[FSDP2] Add cache for FSDP wrapper class

yf225 opened the pull request:
Currently, `fully_shard` will create a new `FSDPMyModuleClass` class for each `MyModuleClass` module **object**, which causes Dynamo to guard-fail on every module object's type checking. This PR fixes the issue by caching and reusing previously created FSDP wrapper class.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134135



cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

yf225 committed:
[FSDP2] Add cache for FSDP wrapper class

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -23,6 +23,9 @@
 from ._fsdp_state import _get_module_fsdp_state, FSDPState
 
 
+cls_to_fsdp_wrapper: Dict[Type, Type] = {}
```
nit: wrapper might be a tiny bit misleading
```suggestion
cls_to_fsdp_cls: Dict[Type, Type] = {}
```

awgu reviewed (approved):
LGTM!

yf225 committed:
Update on "[FSDP2] Add cache for FSDP wrapper class"

Currently, `fully_shard` will create a new `FSDPMyModuleClass` class for each `MyModuleClass` module **object**, which causes Dynamo to guard-fail on every module object's type checking. This PR fixes the issue by caching and reusing previously created FSDP wrapper class.




cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

yf225 commented:
@pytorchbot merge

yf225 commented:
@pytorchbot merge -f "don't think it will fail on rocm, skipping to save capacity"

yf225 commented:
@pytorchbot merge -f "don't think it will fail on rocm, skipping to save capacity"
